File: MedianOfMatrix/MedianOfMatrix.cpp
// MedianOfMatrix.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>

std::vector<int> Merge(const std::vector<int>& v, const std::vector<int>& vv)
{
    std::vector<int> result;
    result.reserve(v.size() + vv.size());

    auto i = v.cbegin();
    auto j = vv.cbegin();
    for (; i != v.cend() && j != vv.cend();)
    {
        if (*i < *j)
            result.push_back(*i++);
        else
            result.push_back(*j++);
    }

    for (; i != v.cend(); ++i)
        result.push_back(*i);
    for (; j != vv.cend(); ++j)
        result.push_back(*j);

    return result;
}

std::vector<int> Merge(const std::vector<std::vector<int>>& matrix)
{
    std::vector<int> v(*matrix.cbegin());
    for (auto i = matrix.cbegin()+1; i != matrix.cend(); ++i)
    {
        v = std::move(Merge(v, *i));
    }

    return v;
}

int CalculateMedianofMatrix(std::vector<std::vector<int>>& matrix)
{
    std::vector<int> vec = Merge(matrix);
    return vec[(vec.size()-1) / 2];
}

int main()
{
    while (true)
    {
        int rows = 0;
        int cols = 0;
        std::cout << "The size of the matrix(rows, columns): ";
        std::cin >> rows >> cols;
        if (rows == 0 && cols == 0)
            break;

        std::vector<std::vector<int>> vecOfVec;
        vecOfVec.reserve(rows);

        for (int i = 0; i < rows; ++i)
        {
            std::vector<int> vec(cols);
            std::cout << "The " << i << "th row of the matrix: ";
            for (int i = 0; i < cols; ++i)
                std::cin >> vec[i];

            vecOfVec.push_back(std::move(vec));
        }

        std::cout << "The median element of the matrix is: " << CalculateMedianofMatrix(vecOfVec) << std::endl << std::endl;
    }
}

File: Go/MedianOfMatrix.go
package main

import "fmt"

func Merge(v1 []int, v2 []int) []int {
	v := make([]int, len(v1)+len(v2))
	index := 0
	i := 0
	j := 0
	for i < len(v1) && j < len(v2) {
		if v1[i] < v2[j] {
			v[index] = v1[i]
			index++
			i++
		} else {
			v[index] = v2[j]
			index++
			j++
		}
	}

	for i < len(v1) {
		v[index] = v1[i]
		index++
		i++
	}

	for j < len(v2) {
		v[index] = v2[j]
		index++
		j++
	}

	return v
}

func CalculateMedianOfMatrix(matrix [][]int) int {
	vec := matrix[0]
	for i := 1; i < len(matrix); i++ {
		vec = Merge(vec, matrix[i])
	}

	fmt.Println(vec)
	return vec[len(vec)/2]
}

func main() {
	for {
		fmt.Print("The size of the matrix (rows, colomns): ")
		var rows, cols int
		fmt.Scan(&rows, &cols)
		if rows == 0 && cols == 0 {
			break
		}

		matrix := make([][]int, rows)
		for i := 0; i < rows; i++ {
			matrix[i] = make([]int, cols)
			fmt.Printf("The %dth row: ", i)
			for j := 0; j < cols; j++ {
				fmt.Scan(&matrix[i][j])
			}
		}

		median := CalculateMedianOfMatrix(matrix)
		fmt.Printf("The median of the matrix is %d\n", median)
	}
}
